pkg/util/watch/watch.go

```diff
@@ -172,8 +172,19 @@ func Watch[C Client[L], L metav1.ListMetaAccessor, T any, P Object[T]](
 		return nil, fmt.Errorf("Initial watch failed: %w", err)
 	}
 
+	// Lock the store to pass it into the goroutine, so that we don't have to worry about immediate
+	// operations on the store racing with any deferred additions.
+	store.mutex.Lock()
+
 	// With the successful Watch call underway, we hand off responsibility to a new goroutine.
 	go func() {
+		holdingInitialLock := true
+		defer func() {
+			if holdingInitialLock {
+				store.mutex.Unlock()
+			}
+		}()
+
 		// note: instead of deferring watcher.Stop() directly, wrapping it in an outer function
 		// means that we'll always Stop the most recent watcher.
```

Comment:
I'd just combine the two defers (forcing folks to read the code backwards is just friction that is often best avoided.

Reply:
Going to leave it as-is for now, because I think it makes the high-level control flow easier (IMO: repeatedly seeing "defer" makes it clear that the things are deferred — order doesn't really matter. Whereas combining everything into one function it's easy for your eyes to skip over the leading "defer" at the top)